File: Sponsors/mw-streams-java/src/main/java/com/mwam/hackathon/operators/Distinct.java
package com.mwam.hackathon.operators;

import java.util.HashSet;
import java.util.Set;

/**
 * Select distinct elements from a stream.
 *
 * @param <I> Event type.
 */
public class Distinct<I> extends Operator<I, I> {
    private final Set<I> seen = new HashSet<>();

    public Distinct() {
        super();
    }

    @Override
    public void consume(I input) {
        // add returns false if the item was already in the set
        if (seen.add(input)) {
            emit(input);
        }
    }
}

File: Sponsors/mw_streams_python/mws/operators/distinct.py
from typing import TypeVar, Callable

from mws.operators.operator import Operator


I = TypeVar("I")


class Distinct(Operator[I, I]):
    def __init__(self) -> None:
        super().__init__()
        self.seen: set[I] = set()

    def consume(self, event: I) -> None:
        if event not in self.seen:
            self.seen.add(event)
            self.emit(event)
